ESP32:machine_uart.c: Enable hardware flow control.

robert-hh opened the pull request:
Hardware flow control will be enabled appropriately when rts or cts are defined. The rts-threshold is set 1, resulting in a rxbuf + 1, before rts is raised. Then still FIFO_size - 1 bytes can be received without data loss. The FIFO_size is 128.
Note: Setting rts_threshold to 0 has a strange side effect, that rts pulses with every received byte.

robert-hh committed:
machine_uart.c: Enable hardware flow control

Hardware flow control will be enabled appropriately when rts or cts are defined. The rts-threshold is set 1, resulting in a rxbuf + 1, before rts is raised. Then still FIFO_size - 1 bytes can be received without data loss. The FIFO_size is 128.
Note: Setting rts_threshold to 0 has a strange side effect, that rts pulses with every received byte.

tve commented:
Cool! But not everyone wants flow control enabled ;-) Have you looked at #5567, which is a more complete implementation?

robert-hh commented:
IHMO, hardware flow control and half-duplex are different topics. This PR here just implements what seemed to be forgotten. It enables hardware flow control only if CTS or RTS are defined.  If you define rts, then RTS flow control will be enabled, if you define cts, then CTS flow control will be enabled. The default is, that rts and cts are not defined, and then hardware flow control is disabled. 
No need for an additional keywords, no inherent dependencies between modes like they are in #5567. In that sense this PR is minimal.

tve commented:
I see, I did not understand that you're simply proposing "hey, if you define and RTS pin then you must want HW flow control on that". While I like the minimalism I'm not sure how realistic this position is at this point. When I first used the UART module I was confused by the meaning of `UART_PIN_NO_CHANGE` -- it's not `USART_NO_PIN` or something like that... What does "no change" mean? Does that mean the assignment of the RTS and CTS pins depends on whether there was a prior use? Or does it mean the default pin assignment? I haven't dug into the code...

Note also that #5567 is not just about rs485 but also about flow-control.

IMHO, this PR would need a couple of elements:
- do other ports have flow-control and is this PR in line with them?
- is it reasonable to expect that nothing breaks by this PR, e.g., folks defining an RTS pin "inadvertantly"? I believe I actually did do exactly that in my code because I thought that `UART_PIN_NO_CHANGE` meant to keep the default esp32 pin assignment, not to have no pin assigned...
- is a value of 1 for the flow control threshold the only sensible value?
- docs

robert-hh commented:
I see this PR as a bug fix (or filling an unintended gap) and not the introduction as a new feature. And I have seen that PR for RS485, and it is surely a good one.
About your questions:
> do other ports have flow-control and is this PR in line with them?
The CC3200 port has the same link between RTS/CTS definition and flow control. The flow control for Pyboard you mentions is for the pyb.uart() class, and there the pins to be used cannot be specified. So there must be an option for enabling and disabling flow control. If that method only works for ESP32, it should be documented. There are many properties that are specific to ports.
> is it reasonable to expect that nothing breaks by this PR?
If you define RTS and do not connect it to the opposite sender, then nothing happens. For CTS it might result in not sending data. But, as said, the only reason to define RTS and/or CTS pins is to use them for hardware based (automatic) flow control. If you want to implement other means of flow control, you can use GPIO ports and control them via machine.pin().
> is a value of 1 for the flow control threshold the only sensible value?
The rts_threshold tells, how much reserved space will be left in the FIFO. That is FIFO_Size - rts_threshold. The hardware RTS signal will switch when the ring buffer is full and rts-threshold bytes are in the FIFO. 0 cannot be used, due to a phenomenon which I would consider a bug. So any value between 1 and 127 is fine. 1 is the most conservative one, giving the sender the longest time to react. I considered to introduce another keyword for defining the threshold, but it does not seem urgently needed. If you need more characters to be buffered, you can increase rxbuf.
> docs
That is a single sentence which can easily be added.

robert-hh committed:
machine.UART.rst: Mention the rts and cts keywords

The rts and cts keywords available in the ESP32 ports were not mentioned yet. That's fixed now, in combination with the commit to actually enable hardware flow control and not only naming pins for it.

tve commented:
@robert-hh thanks for responding to my comments, very appreciated. I know you're "just trying to fix a small bug" and some of what I tried to do is to nudge you to do a tad more, e.g. docs :-)

WRT "nothing breaks" the CTS effect is what concerned me, but if CC3200 uses the same emthod then it makes sense.

WRT "threshold=1", if I understand correctly, RTS will switch as soon as one char is in the HW buffer until the ESP-IDF interrupt handler removes it? That seems to be fraught with issues espcially at high baud rates (e.g. megabits/sec) with RTS pulsing constantly, no? That's bound to affect throughtput? Maybe I'm not quite understanding and need to dive into the details?

peterhinch commented:
A threshold of 1 will usually have adverse effects on throughput. One reason for having a buffer is to enable the link to operate during periods when the application code is busy. This is of particular importance in **uasyncio** applications. If the link stalls as soon as the buffer contains one or two characters, the coroutine performing input be descheduled after those few characters are read. As soon as one or two more characters are received the link will be stalled until the coro is rescheduled. This has the potential to cripple throughput.

A system having hardware flow control at both ends of the link will work best if the buffer is allowed to almost fill before the receiver deasserts RTS. This is because hardware flow control can typically be expected to operate within one character period. Setting a very low value for the threshold only benefits systems with a pathologically slow response to RTS.

A more usual approach would be to enable the threshold to be configurable. If it must be set in stone I would suggest something like 75% full.

robert-hh commented:
@tve @peterhinch Actually there are two stacked buffers. One is the buffer in the driver, which size can be specified by the rxbuf option anda default size of 256 bytes. The other is the FIFO in the UART circuitry, which is configured to be 128 bytes. The rts_threshold is about the FIFO. So RTS will be raised, when a total of rxbuf + FIFO_size characters are buffered. With the default of 256 bytes for rxbuf and a value of rts_threshold of 1 it is 257 bytes. But setting rts_threshold to another value like 64 is not very critical. Only the value of 0 and values close to the FIFO size should be avoided. 
Given that this value is not critical and the user can always increase rxbuf, I did not implement the code for another keyword option, even if that is pretty simple.

tve commented:
The interrupt handler is pretty hair-raising complex and the speed at which the HW is emptied depends on this monster, plus any other interrupt activities that may be going on. If I recall correctly, chars then go into a freertos ring buffer before they land in that second "rxbuf". I don't know what the max data rate is, but at 1Mbps 1 char is just 10us, that's not much time for that monster interrupt handler, nor for critical sections that turn interrupts off in other parts of the code.

YMMV will further vary with the app: a test that does nothing else may well run OK with little buffering but start to slow the data rate down a lot in a real app. In the end, the RTS threshold also depends on the device at the other end: if it handles RTS in HW then one can set the threshold to 125 leaving just one or two chars extra, if it handles RTS in SW then more "overrun" may need to be left.

peterhinch commented:
@robert-hh The device at the other end of the link is (obviously) unknown. In general RTS is asserted in an interrupt context and the remote's CTS is handled either in hardware or in a pin interrupt handler. When the threshold is passed RTS is asserted promptly and a typical remote device will quickly stop transmitting in response to RTS false. In this case overrun is minimal; but a truly general solution will also cope with the case where CTS response is slow.

So there are two reasons for having an RX buffer. A (usually minor) reason is to cope with TX overruns as described above. The major reason is to maximise throughput in the presence of latency in the process which empties the buffer. Bear in mind that the reason a system might require RTS control is precisely that, in its absence, the buffer would overflow. High levels of latency are a certainty in systems with RTS control. RX buffer size and threshold level usually need to be designed to optimise throughput.

I lack specialist knowledge of the ESP32 so I'll leave this discussion to you experts. I just wanted to be sure that these issues were considered at a systems level.

robert-hh commented:
Goog mroning Peter. You're up early, it seems. Yes, I know the mechanisms. But I did not dig (yet) into the details of the espressif firmware regarding interrupts. From my tests so far it seems that a speed-optimal value for the rts_threshold is UART_FIFO_LEN - 8, and a more conservative value with a little bit of performance loss is UART_FIFO_LEN - 32. (UART_FIFO_LEN is defined in driver/uart.h as 128).
It surprised me in the tests so far that the FIFO is not cleared in a timely fashion, even if there are constant and fast read requests. I noticed in another project with C-Code, that the espressif UART driver does not immediately send data when told to do so, but has a lag of ~50ms before it starts sending. It could be the same for receive, in that the driver either waits a) for  a timeout or another event like b) FIFO full or c) rts_threshold reached, before it starts emptying the FIFO. There is a uart_set_rx_timeout() which seems to be related to a) , and uart_set_rx_full_threshold() which seems to be related to b).  I did not search for the default settings yet.

robert-hh commented:
I captured the timing of the signals at various baud rates and a rts_threshold of USRT_FIFO_LEN - 32. What is interesting that the rts-deassert signal has always the length of about 10 char times. It is not constant. A sample picture is in the attached link: (https://hidrive.ionos.com/lnk/idYmqORh)

robert-hh commented:
After further tests I come to the conclusion, that hardware flow control does not work. If the receiver is slower than the sender, data gets lost. But that is exacatly the situation flow control is made for. For testing, I configured rts_threshold to 61. Then rts will go to a high level after receiving 64 bytes. I made a sending program which sends data in 64 bye chunks, with a sequence number in the first four bytes.
And I wrote a little receiver, which receives data, prints out the number of bytes and the first four bytes, and then sleeps for 100ms. Baudrate 115200. The overall timing during the test run is here:
https://hidrive.ionos.com/lnk/IPYGKsau
The print-out of the receiver is:
```
64 b'0000'
256 b'0001'
256 b'0007'
192 b'0013'
768 chars received 4 events
```
The first packet is received well, then the receiver start sleeping. At that time, 
the next 6 packets arrive, before rts is de-asserted permanently. 
The first four of them are stored to the rx buffer. The next two are discarded.
The same happens to the next series of six packets. The final three 
packets fit to the receive buffer. So the packets with the numbers 
5, 6, 11, 12 are lost, as the do not fit into the receive buffer, when 
they arrive.
I am using IDF 4.0. I do not know if espressif changed that part of the code since then.
Receiving code:
```
from machine import Pin, UART
import time

rx=4
tx=5
rts=21
cts=22

def run(baud=115200, wait = 100):

    uart=UART(1, baud, bits=8, stop=1, parity=None, rx=rx, tx=tx, cts=cts, rts=rts)
    num = 0
    count = 0
    try:
        while True:
            n = uart.any()
            if n:
                data = uart.read()
                print(n, data[0:4])
                num += n
                count += 1
                time.sleep_ms(wait)
    except KeyboardInterrupt:
        pass
    print (num, "chars received", count, "events")
```

robert-hh committed:
esp32/machine_uart.c: Set rts_threshold to 96

This is 75% of the FIFO size. This setting slows down the serial throughput only by about 7%.

robert-hh commented:
I tested that with various threshold sizes and have set it to UART_FIFO_LEN - 32, which is
75% of the FIFO size. The speed impact of selecting lower values is not linear and does
not really matter, since one uses that mode especially when the receiving side is slow.

dpgeorge commented:
This was implemented in #6902 (see a3675294ae39a724aec6959238c45af0dea92f21) via the `flow` keyword argument, which matches stm32 and rp2.
